### DISCUSSION TITLE
MultiDiGraph 's pickle read issues【netwokx3.1】

### DISCUSSION THREAD

**Discussion by WangYunFly** [networkx/networkx#6721] [Q&A] [answered]
【netwokx3.1】
I want to write and read NetworkX  MultiDiGraph by code:
```python
with open('graph.gpickle', 'wb') as f:
    pickle.dump(G, f, pickle.HIGHEST_PROTOCOL)
with open('graph.gpickle', 'rb') as f:
    G_read = pickle.load(f)
```
However, the MultiDiGraph I read **G_read** is different from what I saved before.
![image](https://github.com/networkx/networkx/assets/54883961/ce2a1b8f-1604-4b92-8c8c-43bc2f62444b)
![image](https://github.com/networkx/networkx/assets/54883961/8877e633-5aac-4563-9dd5-4bb870f07fe9)
Does it miss some attributes? 
When I try to check `G_read.number_of_edges()`, it breaks down with `KeyError: nan` in `networkx\classes\reportviews.py`
at` preds = self._pred[n]`.
How can I save and read NetworkX  MultiDiGraph with no error?
Thanks for your reply!

**Comment by MridulS** [comment]
There is probably some `nan` like value in your graph. A `np.nan` will lead to this issue like this, a quick way to reproduce this:

``` python
In [2]: G = nx.MultiDiGraph()

In [3]: G.add_edge(1, np.nan)
Out[3]: 0

In [4]: with open('graph.gpickle', 'wb') as f:
    ...:     pickle.dump(G, f, pickle.HIGHEST_PROTOCOL)
    ...:

In [5]: with open('graph.gpickle', 'rb') as f:
    ...:     G_read = pickle.load(f)
    ...:

In [6]: G_read.number_of_edges()
KeyError
...
...
KeyError: nan
```

what is the type of data of your nodes? Are all of them non-nan like?

**Reply by MridulS** [reply]
What does the following bit of code outputs for you `set(type(u) for u in G)`?

**Reply by WangYunFly** [reply]
Thanks for your reply, I try to run the code and find some nodes are 'nan' when I construct the graph. 
But I still have a fuse about why after `write`, the graph I `read ` will occurs this problem.

**Reply by dschult** [reply, marked as answer]
`nan` is a floating point number. Floating point numbers have round-off error associated with them, so they are often not good choices as keys in a dict. When you go to lookup the key if there is any round-off error the dict will not be able to find your number.  `nan` stands for "not-a-number" and it has the additional property in Python that it is not equal to any other number including itself.  So, unless you have the same position in memory as the `nan` value that you used as a dict-key, it will not find the key in the dict.  To check this, create a dict with the key nan and then try to look it up.  You will get a key-error:   `{float("nan"):0}[float("nan")]`

That's why it is occurring in your case.  But in general it is not a good idea to use floating point values as nodes in NetworkX. Round-off error is possible. Try replacing the node names with integers or strings.  And definitely figure out why you have nan values and figure out how you will handle those... :}
